test/e2e/framework/job/wait.go

```diff
@@ -82,14 +83,17 @@ func waitForJobPodsInPhase(ctx context.Context, c clientset.Interface, ns, jobNa
 // both conformance CI jobs with GA-only features and e2e CI jobs with all default-enabled features.
 // So, we need to skip "Complete" condition reason verifications in the e2e conformance test cases.
 func WaitForJobComplete(ctx context.Context, c clientset.Interface, ns, jobName string, reason *string, completions int32) error {
-	if err := wait.PollUntilContextTimeout(ctx, framework.Poll, JobTimeout, false, func(ctx context.Context) (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
+	if err := framework.Gomega().Eventually(ctx, func() int32 {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			framework.Failf("getting the job %v/%v: %v", ns, jobName, err)
 		}
-		return curr.Status.Succeeded == completions, nil
-	}); err != nil {
-		return nil
+		if isJobFailed(job) {
+			framework.Failf("job %v/%v failed while waiting for its completion", ns, jobName)
```

Comment:
/lgtm cancel

You cannot use assertions inside a `Eventually` callback: `Eventually` is from gomega and doesn't know how to deal with the panic raised by ginkgo.

For transient errors that should be retried, return that error with wrapping as a second return value.

For permanent errors, use `gomega.StopPolling(...).Now()`.